Proposal: WorkOrder endpoints

dietrichm opened the pull request:
To be able to rebuild the WorkOrder functionality into the new mobile application, we are designing and building new API endpoints to manage the WorkOrders and related data.

A few major distinctions compared to the endpoint in v1 are that we introduce draft WorkOrders, separate endpoints to manage images (usually photos included in the WorkOrder PDF) and make explicit the various actions and some of the side effects (i.e. signing, finalising and sending).

---

The goal of this PR is to discuss WorkOrder API design, **not** to merge this as a whole. We will be gradually implementing and documenting parts (endpoints and/or a subset of provided properties) of this API proposal over several iterations.

Best to review the complete diff at once.

dietrichm committed:
Added first proposal for workOrders.create endpoint

dietrichm committed:
Include Work Orders source file

dietrichm committed:
Recompile apiary.apib

JorenSaey reviewed (dismissed):
It seems that it has everything needed to complete the first flow for adding a draft

JordyDeGroeve commented on src/datastructures.apib:
```diff
@@ -140,3 +140,121 @@
+                    + company
+            + id: `09e5d75f-f817-4872-9723-57fbb8ff710d` (string, required)
+        + address (Address)
++ distance: `25.2` (number, optional)
```
Q: should we pass a unit like we do with currency?
PP: at least add a comment to specify this is in km.

stivni replied:
I like the idea about the units, although I don't think we'll go into imperial stuff soon 🙈
Not sure if it's worth it to always have to provide that, although I like the explicitness of it

dietrichm replied:
Let's add a default fixed unit.

JordyDeGroeve commented on src/datastructures.apib:
```diff
@@ -140,3 +140,121 @@
+                    + contact
+                    + company
+            + id: `09e5d75f-f817-4872-9723-57fbb8ff710d` (string, required)
+        + address (Address)
```
Feels strange to call it `intervention_address` and pass a customer. What was the reasoning behind this?

WouterSioen replied:
There are two ways you can show that the intervention happened at a different address than the one of the customer:

1. it is the address of a second customer (f.e. an overlapping company needs to get the bill for work done on one of their sub-companies)
2. it is a hand-filled address (so not coupled to a customer)

WouterSioen replied:
`Intervention location` might make more sense than `intervention_address` in that case? (still with a customer or an address)

JordyDeGroeve replied:
`intervention location` seems better to me.

stivni replied:
but a contact or a customer can have different addresses... so shouldn't we provide more detail that just the link? and what if the address changes over time? I think storing the address in all cases, makes more sense... Keeping a link to where the address comes from, might make sense

dietrichm replied:
Kept the `intervention_address` property but made the customer a property of the address, so we always have actual and non changing address fields but with an optional link to a customer entity.

stivni commented on src/datastructures.apib:
```diff
@@ -140,3 +140,121 @@
         + timeTracking.updated
         + user.deactivated
+
+## WorkOrderData (object)
```
PP: I actually don't like that complete responses are modelled as datastructures
there is not a lot of reusability in that and it makes you have to check two files all the time
I used datastructures for things that are value-object like things

stivni commented on src/datastructures.apib:
```diff
@@ -140,3 +140,121 @@
+            + contact
+            + company
+    + id: `09e5d75f-f817-4872-9723-57fbb8ff710d` (string, required)
++ datetime: `2017-05-26T10:01:49+00:00` (string, required)
```
Q : what does this `datetime` represent?

dietrichm replied:
It's the date when the work order was performed. I've renamed it to `performed_at`. We will check the naming with product anyway.

stivni commented on src/datastructures.apib:
```diff
@@ -140,3 +140,121 @@
+            + company
+    + id: `09e5d75f-f817-4872-9723-57fbb8ff710d` (string, required)
++ datetime: `2017-05-26T10:01:49+00:00` (string, required)
++ description (string, optional)
```
Q : is this rich text (markdown?)?
should we mention?

dietrichm replied:
For now this is plain text. I added an example description.

stivni commented on src/01-general/work-orders.apib:
```diff
@@ -0,0 +1,199 @@
+
++ Response 204
+
+### workOrders.drafts [GET /workOrders.drafts]
```
PP : I don't think we need a separate endpoint for that

stivni replied:
but in that case we will need to have a "state" field on the workorder and a filter for it

stivni replied:
I just assume we will have the need of showing combined lists of draft and finalized work orders...

dietrichm replied:
To be discussed in meeting.

dietrichm replied:
Distinct endpoint for drafts will go and a state property + filter on `.list` will be added.

stivni commented on src/datastructures.apib:
```diff
@@ -140,3 +140,121 @@
++ department_id: `cef01135-7e51-4f6f-a6eb-6e5e5a885ac7` (string, optional)
++ layout_id: `a05320f6-b3ef-4ff9-a287-2074a12fb3e9` (string, optional)
++ milestone_id: `e94c1363-3426-46a4-921d-fce03f39bbd5` (string, optional)
++ event_id: `9fc14045-5c6c-4ba8-8672-42ea3f26aa63` (string, optional)
```
Q : why do we currently need the event id?

dietrichm replied:
Providing a meeting ID will mark that meeting as done when the WorkOrder is finished.

stivni commented on src/datastructures.apib:
```diff
@@ -140,3 +140,121 @@
+            + id: `09e5d75f-f817-4872-9723-57fbb8ff710d` (string, required)
+        + address (Address)
++ distance: `25.2` (number, optional)
++ users (array, required)
```
Q : are these the "users" of the work order?

dietrichm replied:
These are the employees carrying out the work and tracking time. It might make sense to rename this to `employees` though.

Update: renamed to assignees as per @jenssegers' suggestion.

stivni commented on src/datastructures.apib:
```diff
@@ -140,3 +140,121 @@
++ users (array, required)
+    + (object)
+        + user_id: `4ec55a8c-2d80-472a-a9c2-5ff5ee7eac6a` (string, required)
+        + hourly_rate: `65` (number, optional)
```
B : a rate is money, so should be amount & currency

stivni commented on src/datastructures.apib:
```diff
@@ -140,3 +140,121 @@
+        + hourly_rate: `65` (number, optional)
+        + work_hours (array, optional)
+            + (object)
+                + One Of
```
Q : do we need to provide both ways?

dietrichm replied:
@JorenSaey what do you prefer here? Do you use a reusable component for creating and editing time trackings that supports both formats, or does one format suffice?

JorenSaey replied:
I prefer just the sarted_at and duration combo, that is what we use in our TimeTracking module. :)

stivni commented on src/datastructures.apib:
```diff
@@ -140,3 +140,121 @@
+                        + started_at: `2017-04-26T10:01:49+00:00` (string, required)
+                        + ended_at: `2017-04-26T16:02:00+00:00` (string, required)
+                + description (string, optional)
++ line_items (array, optional)
```
Q: is this really "line items"?
Aren't these the "used products" or something?

dietrichm replied:
Correct. Don't recall why we went for line items anymore though. Likely to keep this in sync with how this is called in the invoicing domain.

dietrichm replied:
To be checked with product for a suitable name.

stivni commented on src/datastructures.apib:
```diff
@@ -140,3 +140,121 @@
+        + product_id: `e2314517-3cab-4aa9-8471-450e73449040` (string, required)
+        + quantity: `3` (number, optional)
+        + description: `An awesome product` (string, optional)
+        + unit_price (object, optional)
```
Q : I'm not sure if this API should care about prices?

stivni replied:
Same actually for descriptions etc

dietrichm replied:
It's possible a custom price is being used for an existing product. E.g. when a contractor is getting the product on short notice from a different supplier.

dietrichm replied:
Agreed to remove this from the proposal for now. Will be revisited later.

stivni commented on src/01-general/work-orders.apib:
```diff
@@ -0,0 +1,199 @@
+
+    + Attributes (object)
+        + id: `4afb0a9c-91c6-49ed-a2e5-ce7c1e3a87fb` (string)
+        + Include WorkOrderData
```
PP : I'm not sure if we should allow updating all info on the workorder

jenssegers replied:
Following Stijn on this, not the biggest fan of the embedded data structures from the start too 🙈

dietrichm replied:
Since we have draft WorkOrders that can be very minimal it does make sense to allow updating most properties on an update, perhaps except the required ones such as customer.

Regarding the embedded data structures, I inlined those on my local branch since it's indeed a hassle when properties (sets) are different on different endpoints.

stivni commented on src/01-general/work-orders.apib:
```diff
@@ -0,0 +1,199 @@
++ Response 200 (application/json;charset=utf-8)
+
+    + Attributes (object)
+        + data (object)
```
Q: Why would we return an id for that image? Is there such a thing as a workorderImage?
Shouldn't a list of image be part of the workorder detail?

dietrichm replied:
Agreeing on the latter, but we do need IDs to be able to delete images from draft WorkOrders.

stivni commented on src/01-general/work-orders.apib:
```diff
@@ -0,0 +1,199 @@
+            + type: `workOrderImage` (string)
+            + id: `a856642a-a476-4b69-8e42-e7d1badb2ce9` (string)
+
+### workOrders.listImages [GET /workOrders.listImages]
```
PP : see previous comment, should be part of the details imo

stivni commented on src/01-general/work-orders.apib:
```diff
@@ -0,0 +1,199 @@
+
+    + Attributes (object)
+        + id: `4afb0a9c-91c6-49ed-a2e5-ce7c1e3a87fb` (string, required)
+        + signature_data: `data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAIAAACQkWg2AAAAAXNSR0IArs4c6QAAAVlpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8AAAASUVORK5CYII=` (string, required) - PNG image, base64 encoded
```
PP: `signature`

stivni commented on src/01-general/work-orders.apib:
```diff
@@ -0,0 +1,199 @@
+
++ Request (application/json;charset=utf-8)
+
+    + Attributes (object)
```
Q : maybe we have to provide sending details here too? who to send it to, where to send duplicates, which layout to use, ...

dietrichm replied:
Additional properties will become obvious when we groom the relevant stories. Going to leave it like this for now.

stivni reviewed (dismissed):
Wow! Good job!

I have however a lot of remarks, stuff to discuss.

A more global main concern is, that we probably won't be able to publish this completely as described here.

I would recommend to make smaller PRs and deliver this step by step, in sync with what the mobile app wants to deliver.

stivni commented:
@dietrichm I only noticed the not about not to merge now, so ignore that part of my comment

jenssegers commented on apiary.apib:
```diff
@@ -3250,6 +3250,208 @@ Update the current timer. Only possible if there is a timer running.
+
++ Response 204
+
+### workOrders.drafts [GET /workOrders.drafts]
```
Q: Do we need a split between draft and finalized work orders?

dietrichm replied:
Fixed.

dietrichm committed:
Denote distances are provided in kilometers

dietrichm committed:
Make customer a property of intervention address

This way we always refer to an intervention address but allow that
address to be linked to a customer.

Now we don't risk losing the actual work order address when the address
of the customer changes. Also, customers can have multiple addresses
which is not possible to select using the original format.

dietrichm committed:
Inline WorkOrderData in endpoints

dietrichm committed:
Inline WorkOrderInfo and WorkOrderInfoDetail in endpoints

jenssegers commented on apiary.apib:
```diff
@@ -3519,3 +3721,121 @@ Unregister a webhook.
+            + contact
+            + company
+    + id: `09e5d75f-f817-4872-9723-57fbb8ff710d` (string, required)
++ datetime: `2017-05-26T10:01:49+00:00` (string, required)
```
Should be something that ends in `_at`

jenssegers commented on apiary.apib:
```diff
@@ -3519,3 +3721,121 @@ Unregister a webhook.
++ layout_id: `a05320f6-b3ef-4ff9-a287-2074a12fb3e9` (string, optional)
++ milestone_id: `e94c1363-3426-46a4-921d-fce03f39bbd5` (string, optional)
++ event_id: `9fc14045-5c6c-4ba8-8672-42ea3f26aa63` (string, optional)
++ intervention_address (object, optional)
```
Location?

dietrichm replied:
See https://github.com/teamleadercrm/api/pull/290#discussion_r255568364

jenssegers commented on apiary.apib:
```diff
@@ -3519,3 +3721,121 @@ Unregister a webhook.
+                    + company
+            + id: `09e5d75f-f817-4872-9723-57fbb8ff710d` (string, required)
+        + address (Address)
++ distance: `25.2` (number, optional)
```
Distance of what to where? Driving distance?

jenssegers commented on apiary.apib:
```diff
@@ -3519,3 +3721,121 @@ Unregister a webhook.
+            + id: `09e5d75f-f817-4872-9723-57fbb8ff710d` (string, required)
+        + address (Address)
++ distance: `25.2` (number, optional)
++ users (array, required)
```
Would try to find a fitting name inside the workorder domain that represents the people that will work or worked on this. Wikipedia seems to mention something like assignees.

jenssegers commented on apiary.apib:
```diff
@@ -3519,3 +3721,121 @@ Unregister a webhook.
+                        + started_at: `2017-04-26T10:01:49+00:00` (string, required)
+                        + ended_at: `2017-04-26T16:02:00+00:00` (string, required)
+                + description (string, optional)
++ line_items (array, optional)
```
Does this represent the jobs to be performed?

JordyDeGroeve replied:
Yes but mostly products, I assume.
Something clearer than `line_items` would be nice though

dietrichm replied:
Renamed to `products`.

dietrichm committed:
Add example WorkOrder description

dietrichm committed:
Remove distinct workOrders.drafts endpoint

We now have but one .list endpoint with an optional state filter.

dietrichm committed:
Let WorkOrders have a state property that is either draft or finalized

dietrichm committed:
Allow filtering the list endpoint on WorkOrder state

dietrichm committed:
Fix the customer filtering to be optional

dietrichm committed:
Make explicit what customer property is for on intervention address

dietrichm committed:
Specify distance as object with type (currently only kilometres)

dietrichm committed:
Remove redundant specification on .info description

dietrichm committed:
Rename users to more correct domain term assignees

dietrichm committed:
Remove product (line item) properties for custom product entries

This is currently not in scope and we will revisit this when necessary.
Adding products to be persisted will happen through products.add.

dietrichm committed:
Rename line_items collection to products

dietrichm committed:
Allow but one set of properties to create work hours

Only started_at and duration properties are taken.

dietrichm committed:
Include images in .info endpoint instead of distinct .listImages

dietrichm committed:
Remove _data suffix from base64 encoded string propertiy names

dietrichm committed:
Rename datetime to performed_at

dietrichm committed:
Rename distance property to distance_travelled

dietrichm committed:
Make hourly rate use the Money data structure

dietrichm committed:
Add download endpoint for work order PDFs

dietrichm committed:
Recompile apiary.apib

WouterSioen commented on apiary.apib:
```diff
@@ -3250,6 +3250,357 @@ Update the current timer. Only possible if there is a timer running.
+            + id: `09e5d75f-f817-4872-9723-57fbb8ff710d` (string, required)
+        + performed_at: `2017-05-26T10:01:49+00:00` (string, required)
+        + description: `Pipework intervention` (string, optional)
+        + department_id: `cef01135-7e51-4f6f-a6eb-6e5e5a885ac7` (string, optional)
```
Q: is this really optional?

dietrichm replied:
It is currently optional in the client API. We can fall back to the default department or leave the entity's property at `null` or `-1` as is currently the case when it isn't provided.

WouterSioen commented on apiary.apib:
```diff
@@ -3250,6 +3250,357 @@ Update the current timer. Only possible if there is a timer running.
+                    + company
+            + id: `09e5d75f-f817-4872-9723-57fbb8ff710d` (string, required)
+        + performed_at: `2017-05-26T10:01:49+00:00` (string, required)
+        + description: `Pipework intervention` (string, optional)
```
Q: can we keep this optional when we're starting to build an overview of workorders somewhere? I guess we'll need this to show which workorder is which?

dietrichm replied:
Good question. I think this will have to be checked with UX and the Mobile team. Anyway the description is optional in the current client API.

@JorenSaey do you have wireframes and if so what is specified on the WorkOrder overview?

JorenSaey replied:
I think we can do the same thing like TimeTracking: display it like (no description) in the ListView. If it's optional now, the next api should perhaps also make it optional. My opinion is that it is not for front end to decide whether something is required or not. We have other ways to show this ;)

dietrichm replied:
To be checked. 👍

WouterSioen commented on apiary.apib:
```diff
@@ -3250,6 +3250,357 @@ Update the current timer. Only possible if there is a timer running.
+        + milestone_id: `e94c1363-3426-46a4-921d-fce03f39bbd5` (string, optional)
+        + event_id: `9fc14045-5c6c-4ba8-8672-42ea3f26aa63` (string, optional)
+        + intervention_address (object, optional)
+            + Include Address
```
This should also be optional I guess? Can't we use an `OR` here?

dietrichm replied:
I based this on feedback (https://github.com/teamleadercrm/api/pull/290#discussion_r255459660) from @stivni. Right now you have to _always_ provide address fields, so the address stays untouched on the WorkOrder entity (even when any of the customer's addresses you used changes over time).

You _can_ however make an optional link to a specific customer to make explicit that is an address of given customer.

jenssegers replied:
Looking at some online work order templates I find things like "job location", "location", "client address"

dietrichm replied:
This now renamed to `location` and the customer link is removed.

WouterSioen commented on apiary.apib:
```diff
@@ -3250,6 +3250,357 @@ Update the current timer. Only possible if there is a timer running.
+            + type: `km` (enum[string], required)
+                + Members
+                    + km
+            + value: `25.2` (number, required)
```
Is this the total distance travelled? Or is this the average per assignee?

dietrichm replied:
Product specifies this as the distance from the contractor's main address to the intervention address. They do not take into account separate distances per employee/assignee.

jenssegers commented on apiary.apib:
```diff
@@ -3250,6 +3250,357 @@ Update the current timer. Only possible if there is a timer running.
+                        + contact
+                        + company
+                + id: `09e5d75f-f817-4872-9723-57fbb8ff710d` (string, required)
+        + distance_travelled (object, optional)
```
Same here, this suggests that the work order is already done. Or is this something you fill in on completion? If so I don't think it should be part of draft.

JordyDeGroeve replied:
WO are made on the job, no?
So this seems fine in that case.

jenssegers replied:
From what I understand about work orders is that they are made up front and assigned to someone. And after the work is done, it is completed with more details. From that perspective, it's weird to have `distance_travelled` when the job is not started/completed yet.

jenssegers replied:
Just checked with Ghislaine and a work order can be created up front, and you can optionally specify the distance you will travel to the location. And once you've done the job you can update that number with the amount you actually traveled.

dietrichm replied:
This is renamed to `distance`.

jenssegers commented on apiary.apib:
```diff
@@ -3250,6 +3250,357 @@ Update the current timer. Only possible if there is a timer running.
+                        + company
+                + id: `09e5d75f-f817-4872-9723-57fbb8ff710d` (string, required)
+        + distance_travelled (object, optional)
+            + type: `km` (enum[string], required)
```
PP: `unit`

jenssegers commented on apiary.apib:
```diff
@@ -3250,6 +3250,357 @@ Update the current timer. Only possible if there is a timer running.
+                        + started_at: `2017-04-26T10:01:49+00:00` (string, required)
+                        + duration: 3600 (number, required) - In seconds
+                        + description (string, optional)
+        + products (array, optional)
```
I guess some customers also have services as "products". But not sure if there's a good name that combines something like material/services.

Online templates often have different templates for companies that perform purely services, a combination of materials and services or simply a "description" of the work that is to be done or was done.

jenssegers commented on apiary.apib:
```diff
@@ -3250,6 +3250,357 @@ Update the current timer. Only possible if there is a timer running.
+                    + contact
+                    + company
+            + id: `09e5d75f-f817-4872-9723-57fbb8ff710d` (string, required)
+        + performed_at: `2017-05-26T10:01:49+00:00` (string, required)
```
Does this represent the date on which it should be performed or is this the date the person who performed it fills in when he did it? I think our work orders usually represent work to be done in the future? I think our properties should also reflect that.

Looking at some online work order templates I find things like "expected start date", "starting date", "scheduled start". Often in combination with the actual date on completion.

If it is the completion date it should not be part of the draft I think.

dietrichm replied:
This is renamed to `starts_at`.

jenssegers commented on apiary.apib:
```diff
@@ -3250,6 +3250,357 @@ Update the current timer. Only possible if there is a timer running.
+
+## Work Orders [/workOrders]
+
+### workOrders.draft [POST /workOrders.draft]
```
After reading a bit more about work orders I feel like this can also be `.schedule`. Eg. you schedule a work order, and complete it with details of the work you actually performed.

Possibly in combination with an `.assign` endpoint where you assign the scheduled work order to someone.

jenssegers replied:
After short discussion, it seems that draft is ok.

dietrichm committed:
Merge branch 'master' into workorders-endpoints

dietrichm committed:
Rename type property on distance_travelled to unit

dietrichm committed:
Add sent as a possible state of a Work Order

dietrichm committed:
Recompile apiary.apib

dietrichm commented:
Some property naming will still be checked with Product. However we also still have some room to discuss smaller things in the smaller API doc PRs to follow.

dietrichm committed:
Make WorkOrder ID on update endpoint required

dietrichm committed:
Ensure properties can be left out or set to null on update

dietrichm committed:
Allow creating Work Orders without assignees

dietrichm committed:
Recompile apiary.apib

dietrichm commented on src/01-general/work-orders.apib:
```diff
@@ -61,8 +61,8 @@ Update a draft work order.
     + Attributes (object)
-        + id: `4afb0a9c-91c6-49ed-a2e5-ce7c1e3a87fb` (string)
         + customer (object, required)
+        + id: `4afb0a9c-91c6-49ed-a2e5-ce7c1e3a87fb` (string, required)
```
Don't mind this line moving, it's from editing a hunk manually. Fixed in next commit.

dietrichm committed:
Rename intervention_address property to location

dietrichm committed:
Remove the optional customer link in a location address

There is no agreement on the name and what is exactly a customer on a
location address.

dietrichm committed:
Rename distance_travelled to distance

dietrichm committed:
Rename performed_at to starts_at

dietrichm committed:
Add created_at and updated_at datetimes

dietrichm committed:
Remove hourly rates from assignees

We will evaluate later whether this is needed and the right solution.

dietrichm committed:
Define state as an actual enum of strings

This fixes the rendered documentation.

dietrichm committed:
Replace assignees with an array of work hours having a user reference

Since we removed hourly_rate there was an unnecessary structure of
assignees. We can just accept and return an array of time trackings
having a user reference.

dietrichm committed:
Recompile apiary.apib

psrpinto commented on apiary.apib:
```diff
@@ -3281,6 +3281,327 @@ Update the current timer. Only possible if there is a timer running.
+        + milestone_id: `e94c1363-3426-46a4-921d-fce03f39bbd5` (string, optional)
+        + event_id: `9fc14045-5c6c-4ba8-8672-42ea3f26aa63` (string, optional)
+        + location (Address, optional)
+        + distance (object, optional)
```
Q: Should we define the `distance` object in `datastructures.apib`?

dietrichm replied:
Extracted it. There might be more possible data structures though.

psrpinto replied:
~Would it make sense to add `duration` to the `datastructures.apib`? I considered it but went with the safer option (not make it global) but since we're adding it here too, might make sense to make it a shared data structure.~

EDIT: Ignore, misread this thread

WouterSioen commented on apiary.apib:
```diff
@@ -3281,6 +3281,327 @@ Update the current timer. Only possible if there is a timer running.
+        + products (array, optional)
+            + (object)
+                + product_id: `e2314517-3cab-4aa9-8471-450e73449040` (string, required)
+                + quantity: `3` (number, optional)
```
Q: do we want this to be optional? Isn't it more logical to require this instead of magically defaulting to 1 (I guess?)

dietrichm replied:
Yup, should be required. Changing.

WouterSioen commented on apiary.apib:
```diff
@@ -3281,6 +3281,327 @@ Update the current timer. Only possible if there is a timer running.
+            + (object)
+                + user_id: `4ec55a8c-2d80-472a-a9c2-5ff5ee7eac6a` (string, required)
+                + started_at: `2017-04-26T10:01:49+00:00` (string, required)
+                + duration: 3600 (number, required) - In seconds
```
Should we make this a VO with a unit? see https://github.com/teamleadercrm/api/pull/299

dietrichm committed:
Make quantity a required property for a WorkOrder product

dietrichm committed:
Introduce a value object to represent work hours duration

dietrichm committed:
Extract distance VO into a reusable data structure

dietrichm committed:
Extract duration VO to a data structure

dietrichm committed:
Recompile apiary.apib

psrpinto reviewed (dismissed):
LGTM

WouterSioen commented on apiary.apib:
```diff
@@ -3281,6 +3281,311 @@ Update the current timer. Only possible if there is a timer running.
+            + id: `09e5d75f-f817-4872-9723-57fbb8ff710d` (string, required)
+        + starts_at: `2017-05-26T10:01:49+00:00` (string, optional)
+        + description: `Pipework intervention` (string, optional, nullable)
+        + department_id: `cef01135-7e51-4f6f-a6eb-6e5e5a885ac7` (string, optional, nullable)
```
Q: do we want to be able to pass `null` as department id? I guess that if we draft a work order and we don't pass a department, the default department will be taken, and we don't want to be able to remove that one again?

dietrichm replied:
We actually allow the same for deals.update and do allow to "unset" the department again. Perhaps if a work order was drafted for a specific department and the client wants it to be linked to the default department instead (without knowing which one that is), it is possible by setting `department_id` to `null`.

WouterSioen replied:
Okay, that also seems to be the same way it is done in API v1

WouterSioen commented on apiary.apib:
```diff
@@ -3281,6 +3281,311 @@ Update the current timer. Only possible if there is a timer running.
+        + starts_at: `2017-05-26T10:01:49+00:00` (string, optional)
+        + description: `Pipework intervention` (string, optional, nullable)
+        + department_id: `cef01135-7e51-4f6f-a6eb-6e5e5a885ac7` (string, optional, nullable)
+        + layout_id: `a05320f6-b3ef-4ff9-a287-2074a12fb3e9` (string, optional, nullable)
```
Q: do we want to be able to pass `null` as layout id? I guess that if we draft a work order and we don't pass a layout, the default layout for the current department will be taken, and we don't want to be able to remove that one again?

dietrichm replied:
I'd use the same approach here as for the department ID.

dietrichm committed:
Rename products to materials

This is a more common name in the domain of work orders.

dietrichm committed:
Recompile apiary.apib

dietrichm commented:
Since we've been implementing parts and adding separate PRs that have been merged already, the conflicts are expected and will not be resolved. We will never merge this PR.

psrpinto reviewed (dismissed):
Conflict alert

jenssegers commented:
Going to remove myself as reviewer since this PR is not meant to be merged.

dietrichm committed:
Specify work hours duration again as a plain number

This way we keep it in sync with the time tracking endpoints even though
the tasks endpoint introduced the duration object with a unit.

dietrichm committed:
Adapt info endpoint description to one from split PR

dietrichm committed:
Don't wrap quantity number in backticks

dietrichm committed:
Recompile apiary.apib

JordyDeGroeve reviewed (dismissed):
Still not meant to be merged, but will approve the changes

dietrichm committed:
Rename work_hours to time

Now it's in sync with the time & materials domain term for work orders.

dietrichm committed:
Do not allow collections to be nullable in update endpoint

dietrichm committed:
Recompile Apiary file

dietrichm committed:
Provide WorkOrder images using temporary download links

This was concluded from a technical analysis following a concern of the
mobile team that the response of a .info call would be pretty heavy with
multiple full resolution images added to a WorkOrder.

By offering them though a temporary downloads link, they can be lazy
loaded and this will result in quicker feedback times in the app.

Later on we will be able to return multiple variations of an image (i.e.
thumbnails) by having multiple images with the same `id` but different
`variation`.

dietrichm committed:
Return images only when the includes parameter was given

We don't want to generate temporary image download URLs for very .info
call.

dietrichm committed:
Recompile apiary.apib

WouterSioen commented on apiary.apib:
```diff
@@ -3487,10 +3487,11 @@ Get details for a single work order.
-            + images (array)
+            + images (array) - Only included with request parameter `includes=images`.
                 + (object)
                     + id: `a856642a-a476-4b69-8e42-e7d1badb2ce9` (string)
```
Q: should we nest this one deeper for variations of the same images e.g. thumbnails?

dietrichm replied:
You mean `id` as a key in a new object, having an array of multiple objects with separate variations? That is possible but we mentioned this during the TA and @stivni made the remark we can perfectly return multiple objects with the same `id` but a different `variation`.

dietrichm replied:
Let's postpone this because we don't know if and when we will add variations (and sample base64 images).

tiagobastos commented on apiary.apib:
```diff
@@ -3281,6 +3281,312 @@ Update the current timer. Only possible if there is a timer running.
+                + (object)
+                    + id: `a856642a-a476-4b69-8e42-e7d1badb2ce9` (string)
+                    + location: `https://cdn.teamleader.eu/file` (string) - A temporary URL where the image can be downloaded
+                    + expires: `2018-02-05T16:44:33+00:00` (string) - Expiration time of the temporary download URL
```
PP: Add a expires after the `created_at` of the WorkOrder.

JordyDeGroeve replied:
What do you mean? A workOrder does not expire. It's the image link that expires.

tiagobastos replied:
Not clear, sorry. 
The date on the `expires: 2018-02-05T16:44:33+00:00` is before the date the WorkOrder was created `created_at: 2019-05-09T11:25:11+00:00`. Should be after no? 
But, again, just a PP.

JordyDeGroeve replied:
Aha! I will make sure this is fixed in the actual implementation.
This PR will never get merged.

WouterSioen commented on apiary.apib:
```diff
@@ -3281,6 +3281,312 @@ Update the current timer. Only possible if there is a timer running.
+
+    + Attributes (object)
+        + id: `4afb0a9c-91c6-49ed-a2e5-ce7c1e3a87fb` (string, required)
+        + image_id: `a856642a-a476-4b69-8e42-e7d1badb2ce9` (string, required)
```
Q: isn't one id enough? (the id of the work order image?)

dietrichm replied:
If we keep both it's clear that you can't pass any file ID there and expect it to work. You need an actual WorkOrder ID.

Although it should be clear from the endpoint name as well. So not sure what to decide here.

WouterSioen replied:
Both are okay for me, just seemed like obsolete data, but you're probably never going to have the file id without the work order id anyway

dietrichm committed:
Rename id property in attach and delete WO image request

This way it's obvious this is the WorkOrder ID, not something else.

dietrichm committed:
Rename WorkOrder Image ID to File ID

We don't have separate types for this. It's essentially a file so treat
it as one.

dietrichm committed:
Recompile apiary.apib

WouterSioen committed:
Add signed flag to workOrders.info

WouterSioen committed:
Use id instead of work_order_id for the primary resource

We always use `id` as primary resource id, even when the endpoints uses
two entities. There's always a primary one (as stated in the endpoint
name)

WouterSioen committed:
Sign using the finalize endpoint

We don't want to be able to sign a work order and then still update a
work order. That's why it seems logical to pass a signature while
finalization. This means that finalization will always generate a pdf,
but in one case it will be signed and the other not.

kevinwaelkens commented on apiary.apib:
```diff
@@ -3281,6 +3281,302 @@ Update the current timer. Only possible if there is a timer running.
+                + (enum[string])
+                    + draft
+                    + finalized
+                    + sent
```
Q: Isn't this part incorrect? Renders like this:
![Screenshot 2019-05-21 14 02 21](https://user-images.githubusercontent.com/9056632/58094475-1445a280-7bd1-11e9-89fd-c171200f8837.jpg)

tiagobastos replied:
You can check the https://github.com/teamleadercrm/api/compare/work-orders branch.
That one has the approved change for work orders.
This PR will never get merged by the way.

kevinwaelkens replied:
That's really hard to find IMO... Isn't there a better way to look at this code? There's no PR for that branch open yet right now?

tiagobastos replied:
Not yet. But you can checkout the branch and make a preview apiary.html file.

dietrichm replied:
How Apiary renders is it admittedly weird. Not sure how we can fix this but it's an issue for multiple endpoints.

Since we revived an internal API endpoints repo we could add them there in the future and let the docs be rendered separately. If we have spare time we could/might move the ones in the integration branch already.

dietrichm commented:
To be moved to internal API.
